### PR TITLE
Copy source file to a temp path instead of using it directly

### PR DESCRIPTION
**Pull request by davidwengier**
Fixes https://github.com/dotnet/roslyn/issues/59137

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
I'm a bit confused by the user scenario here -- we don't want it to be updated with the new content? I guess scenarios where you have multiple VS instances open with multiple solutions aren't exactly _uncommon_, and I'd assume this would make something worse in that case?

**Comment by davidwengier** [comment]
We don't want it to be updated with new content until the DLL that is being referenced is also being updated with that new content.

I guess there are two competing scenarios:

1. Without this fix, if you had two VS instances open, use GTD in one to show source from the other, then change source in instance B, don't recompile, switch to instance A, and you'd see incorrect code (newer code that your DLL)
2. With this fix, if you had two VS instances open, use GTD in one to show source from the other, then change source in instance B, recompile, switch to instance A, recompile, and you'd see incorrect code (older code that your DLL)

Scenario 2 seems less likely to me, and I think seeing old code is better than new code - with old code a user would at least try closing and reopening the document I would hope. Seeing new code you might try to start digging in to why "debugging isn't working" or similar.

**Comment by jasonmalinowski** [comment]
@davidwengier I guess have done scenario 2 on occasion, but never scenario 1. Comon case was I'm fixing bugs in Roslyn, and then also compiling Omnisharp or something trying to consume those fixes; the Omnisharp solution may have open source from the Roslyn checkout with breakpoints so I can debug that on that side too.

**Comment by davidwengier** [comment]
@tmat did you experience this issue? Or was it just theoretical.

I don't have a strong feeling about whether this is better, worse, or no net change.

**Comment by tmat** [comment]
It isn't theoretical. 

> 2. With this fix, if you had two VS instances open, use GTD in one to show source from the other, then change source in instance B, recompile, switch to instance A, recompile, and you'd see incorrect code (older code that your DLL)

This won't happen if you invoke GTD again, right? The next GTD should open the right version of the code.
Can we detect that the DLL the source file originated from was updated and refresh the content?

**Comment by davidwengier** [comment]
> This won't happen if you invoke GTD again, right? 

Neither scenario will happen if you invoke GTD again, because we will check the checksum of the document against the PDB and ensure it's correct.

> Can we detect that the DLL the source file originated from was updated and refresh the content?

I will have to check, but my gut assumption is that it would take a build (of the consuming project) for the new DLL to be used, so we'd have to trigger off that, but logically it would work at least.

**Comment by jasonmalinowski** [comment]
> This won't happen if you invoke GTD again, right? The next GTD should open the right version of the code.
Can we detect that the DLL the source file originated from was updated and refresh the content? (but not until the DLL is actually updated vs the source file is updated).

Oh that's a nice idea if we can make that work! And that also better handles the case where you forget to build the first solution before trying to use it in the second. Which I absolutely have never done. Not once. Nope.

> I will have to check, but my gut assumption is that it would take a build (of the consuming project) for the new DLL to be used

So once the DLL being consumed is updated, the VS instance will refresh the reference, since we need to get updated information for completion or whatever. So absolutely if you know which DLL was used to chase to the source file path, you could retrigger then too.

File: src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs [hunk 1]
```diff
@@ -167,11 +167,15 @@ internal sealed class PdbSourceDocumentLoaderService : IPdbSourceDocumentLoaderS
...
 
-        private SourceFileInfo? TryGetOriginalFile(SourceDocument sourceDocument, Encoding encoding, TelemetryMessage telemetry)
+        private SourceFileInfo? TryGetOriginalFile(string tempFilePath, SourceDocument sourceDocument, Encoding encoding, TelemetryMessage telemetry)
         {
             if (File.Exists(sourceDocument.FilePath))
             {
-                var result = LoadSourceFile(sourceDocument.FilePath, sourceDocument, encoding, FeaturesResources.external, ignoreChecksum: false);
+                // Copy the source file to a temp path so it can't change, making it invalid, while open in VS
+                var tempSourcePath = Path.Combine(tempFilePath, Path.GetFileName(sourceDocument.FilePath));
+                IOUtilities.PerformIO(() => File.Copy(sourceDocument.FilePath, tempSourcePath, true));
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs:176]: Do we need/want to mark this read only? Or is it already done somewhere?

File: src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs [hunk 2]
```diff
@@ -167,11 +167,15 @@ internal sealed class PdbSourceDocumentLoaderService : IPdbSourceDocumentLoaderS
...
         }
 
-        private SourceFileInfo? TryGetOriginalFile(SourceDocument sourceDocument, Encoding encoding, TelemetryMessage telemetry)
+        private SourceFileInfo? TryGetOriginalFile(string tempFilePath, SourceDocument sourceDocument, Encoding encoding, TelemetryMessage telemetry)
         {
             if (File.Exists(sourceDocument.FilePath))
             {
-                var result = LoadSourceFile(sourceDocument.FilePath, sourceDocument, encoding, FeaturesResources.external, ignoreChecksum: false);
+                // Copy the source file to a temp path so it can't change, making it invalid, while open in VS
+                var tempSourcePath = Path.Combine(tempFilePath, Path.GetFileName(sourceDocument.FilePath));
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs:175]: If I have two different files with the same file name (but different paths) in my PDB, is this going to stomp over the same file?

Comment (tmat) [reply src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs:175]: Would be good idea to include checksum in the path (subdirectory).

Comment (tmat) [reply src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs:175]: Or recreate the directory structure of the project.

**Pull request closed without merging**